### ISSUE TITLE
How to extract   features of image using djl model

### ISSUE THREAD

**Issue by yanchaoguo** [awslabs/djl#88]
我期望的是模型在中间某层layer的特征输出，通过这些特征我可以计算与其它图片的相似度

**Comment by lanking520** [comment]
你好，你可以使用DJL的Image API来把image转换成 NDArray，可以试试看如下的api
```
var img = ImageFactory.getInstance().fromUrl("https://github.com/pytorch/hub/raw/master/dog.jpg");
img.toNDArray(NDManager.getInstance());
```
For more information, you can find [here](https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/modality/cv/Image.java).

```
import ai.djl.ndarray.NDArray;
import ai.djl.ndarray.NDManager;
import ai.djl.ndarray.types.Shape;
import ai.djl.ndarray.index.NDIndex;
import ai.djl.modality.cv.*;

NDManager manager = NDManager.newBaseManager();
var img = ImageFactory.getInstance().fromUrl("https://github.com/pytorch/hub/raw/master/dog.jpg");
img.toNDArray(manager);
```
可以直接复制粘贴代码到 https://djl.ai/website/demo.html Block Runner 直接跑，跑完之后download template在本地运行。

**Comment by yanchaoguo** [comment]
> 你好，你可以使用DJL的Image API来把image转换成 NDArray，可以试试看如下的api
> 
> ```
> var img = ImageFactory.getInstance().fromUrl("https://github.com/pytorch/hub/raw/master/dog.jpg");
> img.toNDArray(NDManager.getInstance());
> ```
> 
> For more information, you can find [here](https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/modality/cv/Image.java).
> 
> ```
> import ai.djl.ndarray.NDArray;
> import ai.djl.ndarray.NDManager;
> import ai.djl.ndarray.types.Shape;
> import ai.djl.ndarray.index.NDIndex;
> import ai.djl.modality.cv.*;
> 
> NDManager manager = NDManager.newBaseManager();
> var img = ImageFactory.getInstance().fromUrl("https://github.com/pytorch/hub/raw/master/dog.jpg");
> img.toNDArray(manager);
> ```
> 
> 可以直接复制粘贴代码到 https://djl.ai/website/demo.html Block Runner 直接跑，跑完之后download template在本地运行。

不是简单的转ndarray 需要用vgg16或者其它深度学习模型去做特征提取 ，提取的特征将应用在后续的图片搜索服务

**Comment by frankfliu** [comment]
What you need is the predict result. What can do following:

**Comment by lanking520** [comment]
About feature extraction. There are a couple of ways. 

If you have a pretrained model, say vgg16 in MXNet or Torchvision. After loading the weight, you can extract these layers and [convert them into TorchScript(PyTorch)](https://github.com/awslabs/djl/blob/master/docs/pytorch/how_to_convert_your_model_to_torchscript.md). After that, you can load the model in DJL and run inference to extract feature.

Another solution is to build the model from DJL and train here. We already have a simple Pikachu SSD training example, you can try to extend it further to support full SSD. After that, same steps to extract layers and apply for inference. 

Hope these help.

**Comment by yanchaoguo** [comment]
> About feature extraction. There are a couple of ways.
> 
> If you have a pretrained model, say vgg16 in MXNet or Torchvision. After loading the weight, you can extract these layers and convert them into TorchScript(PyTorch), or MXNet Symbol, or TF Saved Model bundle. After that, you can load the model in DJL and run inference to extract feature.
> 
> Another solution is to build the model from DJL and train here. We already have a simple Pikachu SSD training example, you can try to extend it further to support full SSD. After that, same steps to extract layers and apply for inference.
> 
> Hope these help.

有没有例子 使用什么方式抽取layers and 抽取 feature

**Comment by yanchaoguo** [comment]
> What you need is the predict result. You can try the following:
> 
> 1. Load the model
> 2. Create a Predictor<NDList, NDList>
> 3. run Predictor.predict(NDList)
> 4. Save the output NDArray for later use

private NDList predict(NDList ndList)  这是一个私有方法

**Comment by frankfliu** [comment]
You can use following code to get the result:
```
        Criteria<NDList, NDList> criteria =
                Criteria.builder()
                        .setTypes(NDList.class, NDList.class)
                        .optModelUrls("http://localhost/my-vgg16.zip")
                        .optTranslator(new NoopTranslator())
                        .build();

        try (ZooModel<NDList, NDList> model = ModelZoo.loadModel(criteria);
            Predictor<NDList, NDList> predictor = model.newPredictor()) {
                NDList result =  predictor.predict(input);
        }
```
`NoopTranslator` will do preprocess/postprocess and simply pass NDList to the model and return raw NDList as output.

Generic predict() API is a public method:
https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/inference/Predictor.java#L111

**Comment by yanchaoguo** [comment]
> You can use following code to get the result:
> 
> ```
>         Criteria<NDList, NDList> criteria =
>                 Criteria.builder()
>                         .setTypes(NDList.class, NDList.class)
>                         .optModelUrls("http://localhost/my-vgg16.zip")
>                         .optTranslator(new NoopTranslator())
>                         .build();
> 
>         try (ZooModel<NDList, NDList> model = ModelZoo.loadModel(criteria);
>             Predictor<NDList, NDList> predictor = model.newPredictor()) {
>                 NDList result =  predictor.predict(input);
>         }
> ```
> 
> `NoopTranslator` will do no preprocess/postprocess and simply pass NDList to the model and return raw NDList as output.
> 
> Generic predict() API is a public method:
> https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/inference/Predictor.java#L111

NoopTranslator     中的 processOutput方法返回的
NDList包含图片特征信息吗，不应该是probabilities = list.singletonOrThrow().softmax(0)

**Comment by frankfliu** [comment]
Here is object detection example:
https://github.com/awslabs/djl/blob/master/examples/src/main/java/ai/djl/examples/inference/ObjectDetection.java#L55-L69

You can achieve image classification in a similar way using DJL pretrained model.

You can want to use your own model, you can find image classification example here:
https://github.com/awslabs/djl/blob/master/examples/src/main/java/ai/djl/examples/inference/ImageClassification.java

**Comment by yanchaoguo** [comment]
> You can use following code to get the result:
> 
> ```
>         Criteria<NDList, NDList> criteria =
>                 Criteria.builder()
>                         .setTypes(NDList.class, NDList.class)
>                         .optModelUrls("http://localhost/my-vgg16.zip")
>                         .optTranslator(new NoopTranslator())
>                         .build();
> 
>         try (ZooModel<NDList, NDList> model = ModelZoo.loadModel(criteria);
>             Predictor<NDList, NDList> predictor = model.newPredictor()) {
>                 NDList result =  predictor.predict(input);
>         }
> ```
> 
> `NoopTranslator` will do no preprocess/postprocess and simply pass NDList to the model and return raw NDList as output.
> 
> Generic predict() API is a public method:
> https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/inference/Predictor.java#L111


这里我再次描述一下我的问题，我打算通过vgg或则resnet模型提取图片特征，并将深度模型提取的特征进行索引以供检索；目前没有发现djl有合适的方法可以帮组我有效的提取图片的特征向量(数组)
您提供的上述方式(NDList result =  predictor.predict(input);) 这里的result 并不含有图片特征信息 而是预测的分类结果

**Comment by yanchaoguo** [comment]
> A pretrained model contain a `Block`, you have to navigate the block to locate the layer (Block) you are interested, and get parameters (ParameterList) from that Block.
> Here is a pseudo code
> 
> ```
> Model model;
> Block block = model.getBlock();
> ParameterList paramList = block.getParameters();
> NDList list = new NDList();
> for (Parameter param : paramList) {
>     if (param.getName().startsWith("my_layer_...")) {
>         list.add(param.getArray());
>     }
> }
> ```

I don't need model's parameters  ,but want to get vectors of image that model feed forward ,for example fc2 of vgg16

**Comment by yanchaoguo** [comment]
> @yanchaoguo Thanks for your input. I would like to understand more about your model. Please try to answer the following question.
> 
> 1. Do you already have a pretrained model (e.g vgg16 or resnet50 based)? If so, trained in which platform?
> 2. If you answer No to previous question, here is some guidance I may recommend you to try:
> 
> Grab a pretrained model from GluonCV or TorchVision or frrom DJL. You can do some transfer learning on your own dataset. Then remove the last two layers (It's a sequential block, just do remove block operation in sequential block). Or you can use a pretrained model from DJL, we already have in-house vgg16 Object Detection model. You need to retrain the model with your own dataset and do remove last block in your case. Remove the last block untill it reach to your desired one. Then you can use this model to do extraction.
> 
> If you answered yes, please follow the steps for your engine to remove the last few layers until your extraction layer. Then you can simply load this model in DJL to do feature extraction.
> 
> 特征向量提取需要预训练模型，如果没有，可以从DJL，GluonCV，TorchVison等平台选择一个，都有VGG相关的模型。然后你可以对你自己的数据集进行一次迁移学习。之后可以通过删除最后几层直到你想要提取特征的那一层。Frank在上面演示了你如何从DJL获得一个预训练模型，这个模型需要你改装，你可以通过SymbolBlock.removeLast() 来删除你的层直到你所需要的那一个。

**Comment by yanchaoguo** [comment]
> @yanchaoguo Thanks for your input. I would like to understand more about your model. Please try to answer the following question.
> 
> 1. Do you already have a pretrained model (e.g vgg16 or resnet50 based)? If so, trained in which platform?
> 2. If you answer No to previous question, here is some guidance I may recommend you to try:
> 
> Grab a pretrained model from GluonCV or TorchVision or frrom DJL. You can do some transfer learning on your own dataset. Then remove the last two layers (It's a sequential block, just do remove block operation in sequential block). Or you can use a pretrained model from DJL, we already have in-house vgg16 Object Detection model. You need to retrain the model with your own dataset and do remove last block in your case. Remove the last block untill it reach to your desired one. Then you can use this model to do extraction.
> 
> If you answered yes, please follow the steps for your engine to remove the last few layers until your extraction layer. Then you can simply load this model in DJL to do feature extraction.
> 
> 特征向量提取需要预训练模型，如果没有，可以从DJL，GluonCV，TorchVison等平台选择一个，都有VGG相关的模型。然后你可以对你自己的数据集进行一次迁移学习。之后可以通过删除最后几层直到你想要提取特征的那一层。Frank在上面演示了你如何从DJL获得一个预训练模型，这个模型需要你改装，你可以通过SymbolBlock.removeLast() 来删除你的层直到你所需要的那一个。

According to your prompt, I got the result simply; as follows：

`
        ConcurrentHashMap<String, String> criteriaFilter = new ConcurrentHashMap<>();
        criteriaFilter.put("layers", "50");
        criteriaFilter.put("flavor", "v1");
 
        Model model = MxModelZoo.RESNET.loadModel(criteriaFilter, new ProgressBar());
        SequentialBlock newBlock = new SequentialBlock();
        SymbolBlock block = (SymbolBlock) model.getBlock();
        block.removeLastBlock();
        block.removeLastBlock();
        block.removeLastBlock();
        block.removeLastBlock();
        block.removeLastBlock();
        block.removeLastBlock();

        newBlock.add(block);
        newBlock.add(Pool.avgPool2DBlock(new Shape(3,3)));
        newBlock.add(Pool.maxPool2DBlock(new Shape(3,3)));
        newBlock.add(Blocks.batchFlattenBlock());
        model.setBlock(newBlock);
  
        DefaultTrainingConfig config = new DefaultTrainingConfig(Loss.softmaxCrossEntropyLoss()) 
                .addEvaluator(new Accuracy())  
                .optDevices(Device.getDevices(1)) 
                .addTrainingListeners(TrainingListener.Defaults.logging()); 

        BufferedImage input = BufferedImageUtils.fromFile(Paths.get("D:\\tmp\\ss.jpg"));
        try (Trainer trainer = model.newTrainer(config)) {
          
            NDManager manager = trainer.getManager();
            ParameterStore parameterStore = new ParameterStore(manager, false);
            NDArray array = BufferedImageUtils.toNDArray(manager, input, NDImageUtils.Flag.COLOR);
             
            array = NDImageUtils.toTensor(array);
            array = array.expandDims(0);
             
            NDArray result = newBlock.forward(parameterStore ,new NDList(array), false).head();
            // result shape : (1, 1024)
            System.out.println(new Gson().toJson(result.toFloatArray())); // I need the fixed-length array 
           }
`

**Comment by lanking520** [comment]
That's awesome, we will try to improve this feature further. At least not calling removeLastBlock for a lot of times

**Comment by yanchaoguo** [comment]
> That's awesome, we will try to improve this feature further. At least not calling removeLastBlock for a lot of times

But it didn't work very well，do you have a easier method to make it more rational

**Comment by yanchaoguo** [comment]
> That's awesome, we will try to improve this feature further. At least not calling removeLastBlock for a lot of times

有个问题想请教一下 resnet50最后的全连接层输出的shape是2048 怎么降维到1024 ？

**Comment by lanking520** [comment]
> > That's awesome, we will try to improve this feature further. At least not calling removeLastBlock for a lot of times
> 
> 有个问题想请教一下 resnet50最后的全连接层输出的shape是2048 怎么降维到1024 ？

You can add another fully-connected layer and train for a few epochs. Also there are [a few techniques](kdnuggets.com/2015/05/7-methods-data-dimensionality-reduction.html) you can try on. Presonally I used PCA before, and it helped to keep 85% feature while dimension reduced to 1/3 of the original size. Hope this help

**Comment by frankfliu** [comment]
Fill free to reopen this issue if you have further questions.
